### DISCUSSION TITLE
No way to ignore gun_tunnel_up

### DISCUSSION THREAD

**Discussion by voluntas** [ninenines/gun#322] [Q&A] [answered]
I use Gun as a mechanism to send webhooks, and I write code like `#{tunnel => ProxyStreamRef}` in the gun:post options to support HTTP Proxy.

My current problem is that the process that calls this post always sends `{gun_tunnel_up, ConnPid, StreamRef, Protocol}`.

There is no workaround for this, and I need to write code in the handle_info of the gen_server that calls the post to ignore {gun_tunnel_up, ConnPid, StreamRef, Protocol}}.

Is there any better way to get around this?

**Comment by essen** [comment, marked as answer]
You shouldn't ignore it, it's there to inform you the tunnel is up and that you can send requests. It's sent after you call `gun:connect` or when the SOCKS connection is established. Until you receive this message, any request you send may not end up on the right server, especially when you use `gun:connect`. For example:

* Call `gun:connect`
* Call `gun:get` with the `tunnel` option
* Receive a 500 response from `gun:connect`
* The `gun:get` that was already sent will not reach the remote endpoint but will instead be handled by the proxy as if it was a normal pipeline request
* Confusion

You should treat `gun_tunnel_up` the same way as `gun_up` in your gen_server: it is not ready to do requests until it receives it.

**Reply by voluntas** [reply]
@essen Ah, thank you for your politeness. I noticed it from the replies I received.

Maybe after calling gun:connect/2, I need to wait for {up, Protocol} = gun:await/2.

I had missed this completely.
https://github.com/ninenines/gun/blob/master/src/gun.erl#L757-L758

Thanks again for your quick replies.

**Reply by essen** [reply]
If you are using `gun:await` then you may need to call it multiple times if you connect through proxies: it handles both the `gun_up` and the `gun_tunnel_up` messages. When using `gun:connect` you can mirror what you do with `gun:open` and call `gun:await` right after. If you connect through multiple proxies you can call it after every `gun:connect`. If you were using SOCKS in `gun:open` then you can call it twice in a raw, once for the SOCKS connection and once for the proxied connection.

Glad that the proxy work is being used, enjoy!

**Reply by voluntas** [reply]
Thank you very much for the detailed explanation. We are not using SOCKS at this time.

And I should have read the test code properly.
https://github.com/ninenines/gun/blob/master/test/tunnel_SUITE.erl#L881-L894

I was getting it to work without waiting for {up, Protocol} on gun:await/2, but I can't catch it in case of an error.

Thanks again.

P.S.: The Proxy feature of Gun is working great since HTTP Proxy is often used in Japanese companies.
